testgen/generators.go

```diff
@@ -862,6 +863,140 @@ var EthGetTransactionReceipt = MethodTests{
 	},
 }
 
+func checkBlockReceipts(t *T, hash common.Hash, receipts []*types.Receipt) error {
+	blockReceipts := t.chain.GetReceiptsByHash(hash)
+	if len(receipts) != len(blockReceipts) {
+		return fmt.Errorf("receipts length notmatch(got: %d, want: %d)", len(receipts), len(blockReceipts))
+	}
+	for i := range receipts {
+		got, _ := receipts[i].MarshalBinary()
+		want, _ := blockReceipts[i].MarshalBinary()
+		if !bytes.Equal(got, want) {
+			return fmt.Errorf("receipt mismatch (got: %s, want: %s)", hexutil.Bytes(got), hexutil.Bytes(want))
+		}
+	}
+	return nil
+}
+
+func testBlockReceiptsByNumber(ctx context.Context, t *T, number uint64) error {
+	var receipts []*types.Receipt
+	if err := t.rpc.CallContext(ctx, &receipts, "eth_getBlockReceipts", fmt.Sprintf("0x%x", number)); err != nil {
+		return err
+	}
+
+	block := t.chain.GetBlockByNumber(number)
+	if block == nil {
+		return fmt.Errorf("block not found (number: %d)", number)
+	}
+	hash := block.Hash()
+	return checkBlockReceipts(t, hash, receipts)
+}
+
+func testBlockReceiptsByHash(ctx context.Context, t *T, hash common.Hash) error {
+	var receipts []*types.Receipt
+	if err := t.rpc.CallContext(ctx, &receipts, "eth_getBlockReceipts", hash); err != nil {
+		return err
+	}
+	return checkBlockReceipts(t, hash, receipts)
+}
+
+func testBlockReceiptsByTag(ctx context.Context, t *T, tag string) error {
+	var receipts []*types.Receipt
+	if err := t.rpc.CallContext(ctx, &receipts, "eth_getBlockReceipts", tag); err != nil {
+		return err
+	}
+	var hash common.Hash
+	switch tag {
+	case "earliest":
+		block := t.chain.GetBlockByNumber(0)
+		hash = block.Hash()
+	case "latest":
```

Comment:
Should have `finalized` case.

Reply:
AFAIK, `safe` and `finalized` are only available after the merge, and at least need 1,2 epochs. If those tags are available, we need to generate at least 64 blocks?

Reply:
You're right. It looks like there is an issue with how merge status is determined in geth. Even if `TerminalTotalDifficulty` is set on genesis, it only _actually_ triggers if it receives a message over the engine api. Let me look into this a bit. Even so, it isn't well defined whether explicitly imported blocks (like what happens here with `geth import`) should be considered `safe`, `finalized`, etc.

Reply:
Let me take a look too :)

Reply:
I think we can run geth with jwt secret, and then post a `ForkchoiceUpdatedV2` message to notify the EL to post merge